subsys/bluetooth/services/ras/rreq/ras_rreq.c

```diff
@@ -374,83 +460,119 @@ static uint8_t ras_on_demand_ranging_data_notify_func(struct bt_conn *conn,
 	}
 
-	struct net_buf_simple segment;
+	store_ranging_data_segment(rreq, data, length);
 
-	net_buf_simple_init_with_data(&segment, (uint8_t *)data, length);
-
-	uint8_t segmentation_header = net_buf_simple_pull_u8(&segment);
+	return BT_GATT_ITER_CONTINUE;
+}
 
-	bool first_segment = segmentation_header & BIT(0);
-	bool last_segment = segmentation_header & BIT(1);
-	uint8_t rolling_segment_counter = segmentation_header >> 2;
+static uint8_t ras_real_time_ranging_data_notify_func(struct bt_conn *conn,
+						      struct bt_gatt_subscribe_params *params,
+						      const void *data, uint16_t length)
+{
+	LOG_DBG("Real-time Ranging Data notification received");
 
-	if (first_segment && rolling_segment_counter != 0) {
-		LOG_WRN("On-demand Ranging Data notification received invalid "
-			"rolling_segment_counter %d",
-			rolling_segment_counter);
-		rreq->on_demand_rd.error_status = -EINVAL;
+	if (data == NULL) {
+		LOG_DBG("Ignored Ranging Data notification with no data.");
 		return BT_GATT_ITER_CONTINUE;
 	}
 
-	if (rreq->on_demand_rd.next_expected_segment_counter != rolling_segment_counter) {
-		LOG_WRN("No support for receiving segments out of order, expected counter %d, "
-			"received counter %d",
-			rreq->on_demand_rd.next_expected_segment_counter, rolling_segment_counter);
-		rreq->on_demand_rd.error_status = -ENODATA;
-		return BT_GATT_ITER_CONTINUE;
+	struct bt_ras_rreq *rreq = ras_rreq_find(conn);
+
+	if (rreq == NULL) {
+		LOG_WRN("Real-time ranging data notification received without associated RREQ "
+			"context, unsubscribing");
+		return BT_GATT_ITER_STOP;
 	}
 
-	uint16_t ranging_data_segment_length = segment.len;
+	if (rreq->real_time_rd.data_cb == NULL || rreq->real_time_rd.ranging_data_out == NULL) {
+		LOG_WRN("Ranging data notification received without required buffer "
+			"or callback, unsubscribing");
+		return BT_GATT_ITER_STOP;
+	}
 
-	if (net_buf_simple_tailroom(rreq->on_demand_rd.ranging_data_out) <
-	    ranging_data_segment_length) {
-		LOG_WRN("Ranging data out buffer not large enough for next segment");
-		rreq->on_demand_rd.error_status = -ENOMEM;
+	if (rreq->data_error_status) {
+		/* Already had an error receiving this ranging counter, so exit here. */
 		return BT_GATT_ITER_CONTINUE;
 	}
 
-	uint8_t *ranging_data_segment =
-		net_buf_simple_pull_mem(&segment, ranging_data_segment_length);
-	net_buf_simple_add_mem(rreq->on_demand_rd.ranging_data_out, ranging_data_segment,
-			       ranging_data_segment_length);
-
-	if (last_segment) {
-		rreq->on_demand_rd.last_segment_received = true;
+	if (length < 2) {
+		LOG_WRN("Real-time Ranging Data notification received invalid length");
+		rreq->data_error_status = -EINVAL;
+		return BT_GATT_ITER_CONTINUE;
 	}
 
-	/* Segment counter is between 0-63. */
-	rreq->on_demand_rd.next_expected_segment_counter =
-		(rolling_segment_counter + 1) & BIT_MASK(6);
+	store_ranging_data_segment(rreq, data, length);
+
+	if (rreq->last_segment_received) {
```

Comment:
feedback from Sean: need to consider the case where segments are received in incorrect order, probably need to reset the ranging data out buffer as well